# Exploring eBay Car Sales Data

In this project, we'll work with a dataset of used cars from the classifieds section of the German eBay website.

The data dictionary provided with data is as follows:

`dateCrawled` - When this ad was first crawled. All field-values are taken from this date.

`name` - Name of the car.

`seller` - Whether the seller is private or a dealer.

`offerType` - The type of listing

`price` - The price on the ad to sell the car.

`abtest` - Whether the listing is included in an A/B test.

`vehicleType` - The vehicle Type.

`yearOfRegistration` - The year in which which year the car was first registered.

`gearbox` - The transmission type.

`powerPS` - The power of the car in PS.

`model` - The car model name.

`kilometer` - How many kilometers the car has driven.

`monthOfRegistration` - The month in which which year the car was first 
registered.

`fuelType` - What type of fuel the car uses.

`brand` - The brand of the car.

`notRepairedDamage` - If the car has a damage which is not yet repaired.

`dateCreated` - The date on which the eBay listing was created.

`nrOfPictures` - The number of pictures in the ad.

`postalCode` - The postal code for the location of the vehicle.

`lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the used car listings.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

This dataset mainly consists of strings, some columns have null values and the colums are written in camelcase isntead of the preferred snakecase. First we're going to manually change the names to snakecase.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
col_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

autos.columns = col_names
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning.

Now lets do some basic exploration so we can find out what cleaning needs to be done.

In [7]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

mostly one value: 

need more investigation:registration type

text data that needs to be cleaned: price and odometer

In [8]:
autos.price = autos.price.str.replace("$", "")
autos.price = autos.price.str.replace(",", "")
autos.price = autos.price.str.replace(".", "")
autos.price = autos.price.astype(int)

In [9]:
autos.odometer = autos.odometer.str.replace("km", "")
autos.odometer = autos.odometer.str.replace(",", "")
autos.odometer = autos.odometer.astype(int)

In [10]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [11]:
autos.odometer_km.head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

Now we've edited the price and odometers colums so that they are now integer values.

In [12]:
autos = autos.drop(["num_pictures", "seller", "offer_type"], axis=1)

We'll drop the above columns since the data in all of them is nearly identical.

## Odometer and Price Columns Continued

Now we'll look for data that dosn't look right by analzying the columns further.

In [13]:
autos.price.unique().shape

(2357,)

In [14]:
autos.price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [15]:
autos.price.value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [16]:
autos.price.value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are a large number of listings that have a price that is too low (e.g. 0) or too high (e.g. 99999999). Since eBay is an auction site, we will keep the low values but remove anything above $350,000, since that is when the values jump to unrealistic numbers.

In [17]:
autos = autos[autos["price"].between(1,351000)]

In [18]:
autos.price.describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Now that we've removed the outliers for price, let's look at the odometer column.

In [19]:
autos.odometer_km.describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos.odometer_km.value_counts().sort_index(ascending=False)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64

The odometer prices don't seem to contain any outliers. The numbers are rounded which suggests that sellers had to choose from a set of options.

## Exploring the date columns

First let's find out the distribution of values in the date_crawled, ad_created, and last_seen columns.

In [21]:
autos[["date_crawled", "ad_created", "last_seen", "registration_year"]][0:5]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  
0               2004  
1               1997  
2               2009  
3               2007  
4               2003

In [22]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=False, dropna=False)
        .sort_index()
)

2016-03-05    1230
2016-03-06     682
2016-03-07    1749
2016-03-08    1617
2016-03-09    1607
2016-03-10    1563
2016-03-11    1582
2016-03-12    1793
2016-03-13     761
2016-03-14    1775
2016-03-15    1665
2016-03-16    1438
2016-03-17    1536
2016-03-18     627
2016-03-19    1689
2016-03-20    1840
2016-03-21    1815
2016-03-22    1602
2016-03-23    1565
2016-03-24    1425
2016-03-25    1535
2016-03-26    1564
2016-03-27    1510
2016-03-28    1693
2016-03-29    1656
2016-03-30    1636
2016-03-31    1546
2016-04-01    1636
2016-04-02    1723
2016-04-03    1875
2016-04-04    1772
2016-04-05     636
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

It looks like the data was crawled over a month with a roughly uniform distribution.

In [23]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The data is take from around a year, from June 2015 to April 2016.

In [24]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [25]:
autos.registration_year.describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There are some odd values in the registration year column since the max value is 9999 (it's the 21st century), and the min value is 1000.

In [26]:
autos.registration_year.value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    5
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

Realistically, the earliest registration date could be in the early 1900s. The latest date would be around 2015 since the car can't be registered after the listing was seen.

In [27]:
autos[autos["registration_year"] > 2015]

date_crawled                                               name  \
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
59     2016-03-17 17:50:54               Mercedes_A_Klasse_W_168__A_140_gruen   
65     2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
...                    ...                                                ...   
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   
49919  2016-03-10 09:49:43                                         Fiat_Punto   
49935  2016-04-01 21:48:20  Mercedes_A_klasse_angemeldet_mit_Tuef_und_Auto...   
49938  2016-03-28 18:45:06                     Mercedes_Benz_A_160_Avantgarde   
49991  2016-03-06 15:25:19                                         Kleinwagen   

       price   abtest vehicle_type  registration_year    gearbox  power_ps  \
10       999     test          NaN               2017    manuell        90   
55         1     test          NaN               2017  automatik       224   
59       700  control          NaN               2016    manuell         0   
65       250  control          NaN               2017    manuell        65   
68     10990     test          NaN               2017    manuell       174   
...      ...      ...          ...                ...        ...       ...   
49910  22200     test          NaN               9000  automatik       140   
49919    180     test          NaN               2016    manuell        86   
49935    800     test          NaN               2017  automatik       101   
49938   2300  control          NaN               2016  automatik       102   
49991    500  control          NaN               2016    manuell         0   

          model  odometer_km  registration_month fuel_type          brand  \
10          NaN       150000                   4    benzin     volkswagen   
55     e_klasse       125000                   7    benzin  mercedes_benz   
59     a_klasse       150000                   0    benzin  mercedes_benz   
65       fiesta       125000                   9    benzin           ford   
68      clubman       100000                   0       NaN           mini   
...         ...          ...                 ...       ...            ...   
49910    andere        10000                   3    benzin           opel   
49919     punto       150000                   8       NaN           fiat   
49935  a_klasse       150000                   9       NaN  mercedes_benz   
49938  a_klasse       150000                   6    benzin  mercedes_benz   
49991    twingo       150000                   0    benzin        renault   

      unrepaired_damage           ad_created  postal_code            last_seen  
10                 nein  2016-03-14 00:00:00        86157  2016-04-07 03:16:21  
55                 nein  2016-03-06 00:00:00        22111  2016-03-08 05:45:44  
59                  NaN  2016-03-17 00:00:00        95356  2016-03-19 17:46:47  
65                  NaN  2016-04-04 00:00:00        65606  2016-04-05 12:22:12  
68                 nein  2016-04-03 00:00:00        83135  2016-04-05 17:26:26  
...                 ...                  ...          ...                  ...  
49910               NaN  2016-04-03 00:00:00        56856  2016-04-05 22:18:26  
49919                ja  2016-03-10 00:00:00        59558  2016-03-10 10:39:58  
49935              nein  2016-04-01 00:00:00        39108  2016-04-01 21:48:20  
49938              nein  2016-03-28 00:00:00        13507  2016-04-07 00:44:35  
49991               NaN  2016-03-06 00:00:00        61350  2016-03-06 18:24:19  

[3099 rows x 17 columns]

In [28]:
autos[autos["registration_year"].between(1910, 2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[46681 rows x 17 columns]

In [29]:
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
2006    0.054978
2001    0.054278
2002    0.051189
1998    0.048656
2007    0.046886
Name: registration_year, dtype: float64

It looks like most of the vehicles were registered in the past 20 years.

## Exploring Price by Brand

In [30]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212828
opel              0.108658
bmw               0.108597
mercedes_benz     0.095789
audi              0.085823
ford              0.069639
renault           0.047874
peugeot           0.029445
fiat              0.025986
seat              0.018944
skoda             0.016061
nissan            0.015258
mazda             0.015217
smart             0.014290
citroen           0.014125
toyota            0.012581
hyundai           0.009945
sonstige_autos    0.009698
volvo             0.009039
mini              0.008607
mitsubishi        0.008216
honda             0.007989
kia               0.007104
alfa_romeo        0.006610
porsche           0.005910
suzuki            0.005889
chevrolet         0.005663
chrysler          0.003480
dacia             0.002656
daihatsu          0.002512
jeep              0.002224
subaru            0.002121
land_rover        0.002039
saab              0.001627
daewoo            0.001565
jaguar            0.001524
trabant           0.001400
r

Most of the market share is within a few brands. Most of the brands have less than a 1% market share. Because of this, we will only focus on brands that make up at least 5% of the listings.

In [31]:
top_brands = autos["brand"].value_counts(normalize=True)

In [32]:
print(top_brands)

volkswagen        0.212828
opel              0.108658
bmw               0.108597
mercedes_benz     0.095789
audi              0.085823
ford              0.069639
renault           0.047874
peugeot           0.029445
fiat              0.025986
seat              0.018944
skoda             0.016061
nissan            0.015258
mazda             0.015217
smart             0.014290
citroen           0.014125
toyota            0.012581
hyundai           0.009945
sonstige_autos    0.009698
volvo             0.009039
mini              0.008607
mitsubishi        0.008216
honda             0.007989
kia               0.007104
alfa_romeo        0.006610
porsche           0.005910
suzuki            0.005889
chevrolet         0.005663
chrysler          0.003480
dacia             0.002656
daihatsu          0.002512
jeep              0.002224
subaru            0.002121
land_rover        0.002039
saab              0.001627
daewoo            0.001565
jaguar            0.001524
trabant           0.001400
r

In [33]:
common_brands = top_brands[top_brands > .05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


Only 6 of the brands made up at >= 5% of the listings.

In [34]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5332,
 'opel': 2944,
 'bmw': 8261,
 'mercedes_benz': 8536,
 'audi': 9212,
 'ford': 3728}

Despite being the top brands, they have very different price points. Mercedes, BMW and Audi are the most expensive. Opel and Ford as less expensive, and Volkswagen is in between.

In [35]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

brand_mean_mileage

{'volkswagen': 128896,
 'opel': 129383,
 'bmw': 132682,
 'mercedes_benz': 130796,
 'audi': 129492,
 'ford': 124349}

Now that we've calculated the mean prices and mileage for each of the top brands, now we will convert these dictionaries into dataframes.

In [59]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending=False)
bmp_series

audi             9212
mercedes_benz    8536
bmw              8261
volkswagen       5332
ford             3728
opel             2944
dtype: int64

In [58]:
brand_info = pd.DataFrame(bmp_series, columns=["mean_price"])
brand_info

mean_price
volkswagen           5332
opel                 2944
bmw                  8261
mercedes_benz        8536
audi                 9212
ford                 3728

In [60]:
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
bmm_series

bmw              132682
mercedes_benz    130796
audi             129492
opel             129383
volkswagen       128896
ford             124349
dtype: int64

In [61]:
brand_info["mean_mileage"] = bmm_series

In [62]:
brand_info

mean_price  mean_mileage
volkswagen           5332        128896
opel                 2944        129383
bmw                  8261        132682
mercedes_benz        8536        130796
audi                 9212        129492
ford                 3728        124349

In [65]:
brand_info.sort_values(by=["mean_price"], ascending=False)

mean_price  mean_mileage
audi                 9212        129492
mercedes_benz        8536        130796
bmw                  8261        132682
volkswagen           5332        128896
ford                 3728        124349
opel                 2944        129383